Data/Generate.cpp

```diff
@@ -15,6 +15,99 @@
 namespace ComPWA {
 namespace Data {
 
+std::tuple<std::vector<ComPWA::Event>, double>
+generateBunch(unsigned int EventBunchSize, const ComPWA::Kinematics &Kinematics,
+              ComPWA::Intensity &Intensity,
+              ComPWA::UniformRealNumberGenerator &RandomGenerator,
+              double generationMaxValue,
+              std::vector<ComPWA::Event>::const_iterator PhspStartIterator,
+              std::vector<ComPWA::Event>::const_iterator PhspTrueStartIterator,
+              bool InverseIntensityWeighting) {
+
+  std::vector<ComPWA::Event> SelectedEvents;
+
+  std::vector<ComPWA::DataPoint> tmp_datapoints(EventBunchSize);
+  // evaluate function
+  // Note: some event generators create events outside of the phase space
+  // boundary (due to numerical instability and precision). These events have
+  // to be ignored!
+  std::transform(pstl::execution::par_unseq, PhspTrueStartIterator,
+                 PhspTrueStartIterator + EventBunchSize, tmp_datapoints.begin(),
+                 [&Kinematics](const ComPWA::Event &evt) {
+                   ComPWA::DataPoint point(Kinematics.convert(evt));
+                   if (!Kinematics.isWithinPhaseSpace(point))
+                     point.Weight = 0.0;
+                   return point;
+                 });
+
+  // transform into horizontal data structure
+  std::vector<std::vector<double>> tmp_data;
+  for (auto x : Kinematics.getKinematicVariableNames()) {
+    std::vector<double> tempvec(EventBunchSize);
+    tmp_data.push_back(tempvec);
+  }
+
+  for (size_t i = 0; i < tmp_datapoints.size(); ++i) {
+    for (size_t j = 0; j < tmp_data.size(); ++j) {
+      tmp_data[j][i] = tmp_datapoints[i].KinematicVariableList[j];
```

Comment:
`at()` is just insanely slow, and since at this point you need speed and the boundaries can be checked beforehand, I would avoid using `at()` here. I have no idea how to program this "safer" though.

Reply:
I agree at this point. But in all other parts where performance is not an issue we should use `at()`.

Reply:
I agree, so I can resolve this part?